src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -573,72 +573,86 @@ query_dfs_mount(const char *path)
+/* Discover fuse mount points from /proc/self/mounts and env DAOS_MOUNT_POINT.
+ * Return 0 for success. Otherwise return Linux errno. This function interception
+ * is still effective as long as num_dfs is not zero even if discover_daos_mount_with_env()
+ * returns a non-zero value.
+ */
+static int
+discover_daos_mount_with_env(void)
 {
-	int   idx, len_fs_root;
+	int   idx, len_fs_root, rc;
 	char *fs_root   = NULL;
 	char *pool      = NULL;
 	char *container = NULL;
 
-	/* Find the list of dfuse from /proc/mounts */
-	discover_dfuse();
-
 	/* Add the mount if env DAOS_MOUNT_POINT is set. */
 	fs_root = getenv("DAOS_MOUNT_POINT");
 	if (fs_root == NULL)
-		return;
+		/* env DAOS_MOUNT_POINT is undefined, return success (0) */
+		D_GOTO(out, rc = 0);
 
 	if (num_dfs >= MAX_DAOS_MT) {
-		D_WARN("dfs_list[] is full already. Need to incease MAX_DAOS_MT.\n");
-		return;
+		D_FATAL("dfs_list[] is full already. Need to incease MAX_DAOS_MT.\n");
+		D_GOTO(out, rc = EBUSY);
 	}
 
 	if (access(fs_root, R_OK)) {
-		D_DEBUG(DB_ANY, "no read permission for %s: %d (%s)\n", fs_root, errno,
-			strerror(errno));
-		return;
+		D_FATAL("no read permission for %s: %d (%s)\n", fs_root, errno,	strerror(errno));
+		D_GOTO(out, rc = EACCES);
 	}
 
+	/* check whether fs_root exists in dfs_list[] already. "idx >= 0" means exists. */
 	idx = query_dfs_mount(fs_root);
 	if (idx >= 0)
-		return;
+		D_GOTO(out, rc = 0);
 
 	/* Not found in existing list, then append this new mount point. */
 	len_fs_root = strnlen(fs_root, DFS_MAX_PATH);
 	if (len_fs_root >= DFS_MAX_PATH) {
-		D_DEBUG(DB_ANY, "DAOS_MOUNT_POINT is too long. It is ignored.\n");
-		return;
+		D_FATAL("DAOS_MOUNT_POINT is too long.\n");
+		D_GOTO(out, rc = ENAMETOOLONG);
 	}
 
 	pool = getenv("DAOS_POOL");
 	if (pool == NULL) {
-		D_DEBUG(DB_ANY, "DAOS_POOL is not set.\n");
-		return;
+		D_FATAL("DAOS_POOL is not set.\n");
+		D_GOTO(out, rc = EINVAL);
 	}
 
 	container = getenv("DAOS_CONTAINER");
 	if (container == NULL) {
-		D_DEBUG(DB_ANY, "DAOS_CONTAINER is not set.\n");
-		return;
+		D_FATAL("DAOS_CONTAINER is not set.\n");
+		D_GOTO(out, rc = EINVAL);
 	}
 
 	D_STRNDUP(dfs_list[num_dfs].fs_root, fs_root, len_fs_root);
-	if (dfs_list[num_dfs].fs_root == NULL)
-		return;
+	if (dfs_list[num_dfs].fs_root == NULL) {
+		D_FATAL("D_STRNDUP() failed: %d (%s)\n", ENOMEM, strerror(ENOMEM));
+		D_GOTO(out, rc = ENOMEM);
+	}
 
 	dfs_list[num_dfs].pool         = pool;
 	dfs_list[num_dfs].cont         = container;
 	dfs_list[num_dfs].dfs_dir_hash = NULL;
 	dfs_list[num_dfs].len_fs_root  = len_fs_root;
 	atomic_init(&dfs_list[num_dfs].inited, 0);
 	num_dfs++;
+	rc = 0;
+
+out:
+	return rc;
 }
 
 #define MNT_TYPE_FUSE	"fuse.daos"
-static void
-discover_dfuse(void)
+/* Discover fuse mount points from /proc/self/mounts. Return 0 for success. Otherwise
+ * return Linux errno. This function interception is still effective as long as num_dfs
+ * is not zero even if discover_dfuse_mounts() returns a non-zero value.
```

Comment:
The code in the calling function will not use interception if this function fails which contradicts this comment.

Reply:
Comments are fixed. Thank you!